crypto/modes/siv128.c

```diff
@@ -0,0 +1,365 @@
+/*
+ * Copyright 2017 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <stdlib.h>
+#include <openssl/crypto.h>
+#include <openssl/cmac.h>
+#include "modes_lcl.h"
+
+#ifndef OPENSSL_NO_SIV
+
+# if !defined(BSWAP8)
+#  if defined(__GNUC__) || defined(__clang__)
+__owur static ossl_inline uint64_t BSWAP8(uint64_t x)
+{
+    return __builtin_bswap64(x);
+}
+#  elif defined(_MSC_VER)
+__owur static ossl_inline uint64_t BSWAP8(uint64_t x)
+{
+    return _byteswap_uint64(x);
+}
+#  else
+__owur static ossl_inline uint32_t rotl(uint32_t x)
+{
+    return (x << 8) | (x >> 24);
+}
+__owur static ossl_inline uint32_t rotr(uint32_t x)
+{
+    return (x >> 8) | (x << 24);
+}
+__owur static ossl_inline uint64_t BSWAP8(uint64_t x)
+{
+    uint32_t high = (uint32_t)(x >> 32);
+    uint32_t low = (uint32_t)x;
+
+    high = (rotl(high) & 0x00ff00ff) | (rotr(high) & 0xff00ff00);
+    low = (rotl(low) & 0x00ff00ff) | (rotr(low) & 0xff00ff00);
+    return ((uint64_t)low) << 32 | (uint64_t)high;
+}
+#  endif
+# endif
+
+__owur static ossl_inline uint64_t siv128_getword(SIV_BLOCK const *b, size_t i)
+{
+    const union {
+        long one;
+        char little;
+    } is_endian = { 1 };
+
+    if (is_endian.little)
+        return BSWAP8(b->word[i]);
+    return b->word[i];
+}
+
+static ossl_inline void siv128_putword(SIV_BLOCK *b, size_t i, uint64_t x)
+{
+    const union {
+        long one;
+        char little;
+    } is_endian = { 1 };
+
+    if (is_endian.little)
+        b->word[i] = BSWAP8(x);
+    else
+        b->word[i] = x;
+}
+
+static ossl_inline void siv128_xorblock(SIV_BLOCK *x,
+                                     SIV_BLOCK const *y)
+{
+    x->word[0] ^= y->word[0];
+    x->word[1] ^= y->word[1];
+}
+
+/*
+ * Doubles |b|, which is 16 bytes representing an element
+ * of GF(2**128) modulo the irreducible polynomial
+ * x**128 + x**7 + x**2 + x + 1.
+ * Assumes two's-complement arithmetic
+ */
+static ossl_inline void siv128_dbl(SIV_BLOCK *b)
+{
+    uint64_t high = siv128_getword(b, 0);
+    uint64_t low = siv128_getword(b, 1);
+    uint64_t high_carry = high & (((uint64_t)1) << 63);
+    uint64_t low_carry = low & (((uint64_t)1) << 63);
+    int64_t low_mask = -((int64_t)(high_carry >> 63)) & 0x87;
+    uint64_t high_mask = low_carry >> 63;
+
+    high = (high << 1) | high_mask;
+    low = (low << 1) ^ (uint64_t)low_mask;
+    siv128_putword(b, 0, high);
+    siv128_putword(b, 1, low);
+}
```

Comment:
Question here is how efficient is it in big-endian case. See line #67 in xts128.c...

Reply:
It should have said "in little-endian case", right? It's little-endian in xts, but here it's big-endian. Question basically is what's more efficient in *general* case. Multi-step byte swap or character loop of fixed length. Well, in xts128.c the choice is simpler, because trade-off doesn't apply to most popular platform...

Reply:
It's pretty efficient. The character loop from xts128.c is significantly slower than the implementation here.